File: progonka.py
N = 5 #размерность матрицы или число интервалов + 1
a = 2
y0 = 6
yN = 4
begin = 0
end = 10
h = (end - begin) / (N - 1)

#константы интегрирования для точного решения
C = (y0 - yN - a/2 * (begin*begin - end*end)) / (begin - end)
D = y0 - a/2 * begin*begin - C * begin


def fillSystem():
    f = [y0]
    f.extend([a * h * h for i in range(1, N - 1)])
    f.append(yN)

    A = [[0 for i in range(N)] for i in range(N)]
    A[0][0] = 1
    for i in range(1, N - 1):
        A[i][i - 1] = 1
        A[i][i] = -2
        A[i][i + 1] = 1
    A[-1][-1] = 1

    return A, f


def forward(A, f):
    alpha = [-A[0][1] / A[0][0]]
    beta = [f[0] / A[0][0]]
    for i in range(1, N - 1):
        alpha.append(-A[i][i + 1] / (A[i][i] + alpha[i - 1] * A[i][i - 1]))
        beta.append((f[i] - beta[i - 1] * A[i][i - 1]) / (A[i][i] + alpha[i - 1] * A[i][i - 1]))
    alpha.append(0)
    beta.append((f[-1] - beta[-1] * A[-1][-2]) / A[-1][-1] + alpha[-2] * A[-1][-2])
    return alpha, beta


def backward(alpha, beta):
    x = [0 for i in range(N - 1)]
    x.append(beta[-1])
    for i in range(N - 2, -1, -1):
        x[i] = alpha[i] * x[i + 1] + beta[i]
    return x


A, f = fillSystem()
alpha, beta = forward(A, f)
solve = backward(alpha, beta)

print(solve)
import matplotlib.pyplot as plt
xs = [begin + i * h for i in range(N)]
xe = [begin + i * (end - begin) / 1000 for i in range(1000)]
ys = [a/2 * x*x + C * x + D for x in xe]
ax = plt.axes()
ax.spines['left'].set_position('center')
ax.spines['bottom'].set_position('center')
ax.spines['top'].set_visible(False)
ax.spines['right'].set_visible(False)
xlim = max(abs(begin), abs(end)) + 5
plt.xlim([-xlim, xlim])
ylim = max(abs(min(solve)), abs(max(solve)), abs(max(ys)), abs(min(ys))) + 5
plt.ylim([-ylim, ylim])
plt.plot(xe, ys, 'b', label='exact solution')
plt.plot(xs, solve, 'bo', label='numerical solution')
plt.legend()
plt.show()

File: progonka.c
#include <stdio.h>
#include <stdlib.h>

#define N 101 //размерность матрицы или число интервалов + 1
#define a 2 //решаем уравнение y'' = a
#define y0 0
#define yN 0
#define begin -10.
#define end 10

void fillSystem(double* A, double* f){
	double h = (end - begin) / (N - 1);
	for (int i = 0; i < N; i++)
		A[i] = !i;
	for (int i = 1; i < N - 1; i++){
		for (int j = 0; j < i - 1; j++)
			A[i*N + j] = 0;
		A[i*N + i - 1] = 1;
		A[i*N + i] = -2;
		A[i*N + i + 1] = 1;
		for (int j = i + 2; j < N; j++)
			A[i*N + j] = 0;
	}
	for (int i = 0; i < N; i++)
		A[N * (N - 1) + i] = (i == N - 1);
	f[0] = y0;
	for (int i = 1; i < N - 1; i++)
		f[i] = a * h * h;
	f[N - 1] = yN;
}

void forward(double* A, double* f, double* alpha, double* beta){
	alpha[0] = -A[1] / A[0];
	beta[0] = f[0] / A[0];
	for (int i = 1; i < N - 1; i++){
		alpha[i] = -A[i*N + i + 1] / (A[i*N + i] + alpha[i - 1] * A[i*N + i - 1]);
		beta[i] = (f[i] - beta[i - 1] * A[i*N + i - 1]) / (A[i*N + i] + alpha[i - 1] * A[i*N + i - 1]);
	}
	beta[N - 1] = (f[N - 1] - beta[N - 2] * A[N*N - 2]) / (A[N*N - 1] + alpha[N - 2] * A[N*N - 2]);
}

void backward(double* x, double* alpha, double* beta){
	x[N - 1] = beta[N - 1];
	for (int i = N - 2; i >= 0; i--)
		x[i] = alpha[i] * x[i + 1] + beta[i];
}

int main(){
	double* A = (double*)malloc(N * N * sizeof(double));
	double* x = (double*)malloc(N * sizeof(double));
	double* f = (double*)malloc(N * sizeof(double));
	double* alpha = (double*)malloc(N * sizeof(double));
	double* beta = (double*)malloc(N * sizeof(double));
	fillSystem(A, f);
	forward(A, f, alpha, beta);
	backward(x, alpha, beta);
	for (int i = 0; i < N; i++)
		printf("%lf  ", x[i]);
	free(A);
	free(x);
	free(f);
	free(alpha);
	free(beta);
	return 0;
}
